### ISSUE TITLE
Using python instead of nodejs for grabbing webcam frames

### ISSUE THREAD

**Issue by allo-** [fangfufu/Linux-Fake-Background-Webcam#2]
I read the blog post and saw your project and as I did not want to setup docker and run nodejs, I looked for projects that use bodypix in python and found https://github.com/ajaichemmanam/simple_bodypix_python.

I created a fork and combined their example script with your webcam script and created this repo: https://github.com/allo-/simple_bodypix_python

The program already works, but has a few issues like a worse resolution (I need to investigate how the node module interpolates the mask in such a high resolution).

Would you like to work together?

**Comment by fangfufu** [comment]
Yes, I am happy to help out where I can. I am a bit busy with other stuff, and this is not really the area I normally work in. So I am not sure how much help I can provide.

**Comment by allo-** [comment]
I am rather new to this as well, but maybe we can figure out the missing bits together.

My current issues:

- The mask is rather coarse. Your code uses mobilenet and not resnet, but I do not see what's the right model for the simple_bodypix_python code to use mobilenet.
- On my machine I get a "CuDNN could not be initialized" error that I do not get for other tensorflow projects. This may hint at a too large model (resnet instead of mobilenet) for the GPU?
- I cat watch the webcam with mplayer, but chrome only gives a preview (e.g. in jitsi meet) but does not allow to choose it as input device (Requirements on the framerate?)

**Comment by fangfufu** [comment]
1. I doubt the whole thing can run with real time inference if you use resnet rather than mobilenet. The whole point of mobilenet is that it is fast. 
2. I have no idea. What else does your console output say? 
3. Did you enable the ``exclusive_caps`` on your v4l2loopback device? You also need to start the webcam before starting Chrome.

**Comment by fangfufu** [comment]
Also please refer to:
https://github.com/umlaeute/v4l2loopback/issues/78

I have problem with getting the whole thing working in Chrome as well, it works fine in Skype.

**Comment by allo-** [comment]
1. I think I need to lookup the model architecture differences. Resnet seems to have a different output shape than mobilenet and currently just based the script on the code from simple_bodypix_python.
2. I get the message about the possible initialization error and then an exception that is probably related to this warning. I need to debug this in more detail to see if cuDNN could be loaded or not. There are several possible reasons https://superuser.com/a/1446579 and it may be resolved by using mobilenet.
3. Yes. Otherwise chromium does not detect the device at all.

**Comment by fangfufu** [comment]
With respect to 2), if you have problem with getting this repository working, you could try my cpu-only branch.

Also, please stick to mobilenet, not everyone has an Nvidia GPU. mobilenet is already quite slow running on my quad-core CPU.

**Comment by fangfufu** [comment]
With respect to 3), use Firefox

**Comment by fangfufu** [comment]
Please also refer to 
https://github.com/tensorflow/tfjs/issues/2043

**Comment by allo-** [comment]
On 2: I will need mobilenet anyway. I probably need to have a look into the paper or read the node module source regarding the different tensor shapes in the model.

On 3: Firefox is not an option. WebRTC is incomplete and buggy, so video streaming with tools like jitsi meet does not work very well. This is a known bug* and Mozilla is working on it, but currently the only option is to use a chromium based browser. I think it even drags down the performance for other users, but I am not sure.

When the tensorflow bug about converting to tflite is resolved, I am happy to use the model this way. Currently the script already implements https://github.com/tensorflow/tfjs/issues/2043#issuecomment-602367457

Have a look at my files in the repo, especially the main script:
https://github.com/allo-/simple_bodypix_python/blob/master/fake_webcam_background.py

You need to install `tfjs-to-tf` and `pyfakewebcam` and the packages in requirements.txt. I use latest tensorflow (2.1.0) because I need recent CUDA for other projects, but it should work with TF 1.15.2 as well.

I created three issues for these points in my repo and when they take longer to resolve I will document the progress there. I sent you an invite if you want to contribute, so we can work together on this code. You can ignore the other files for now, I think when the code is more beautiful I will create a new project that only contains the fakewebcam script. Until then it may be useful to be able to test all details with the existing scripts.

* One of the Bugs tracking the progress is https://bugzilla.mozilla.org/show_bug.cgi?id=1164187. I think there are a few other issues. Mozilla and Jitsi try to resolve the issues as soon as possible, but currently Firefox is no good option for video conferencing.

**Comment by allo-** [comment]
Do you know what the last dimension of the mobilenet tensor is?

- Resnet shape: (46, 81)
- MobileNet shape: (46, 81, 32)

**Comment by fangfufu** [comment]
I am not sure, sorry. I don't normally work on neural network.

**Comment by allo-** [comment]
I got it. the order of the output is different, see https://github.com/tensorflow/tfjs-models/blob/master/body-pix/src/mobilenet.ts vs. https://github.com/tensorflow/tfjs-models/blob/master/body-pix/src/resnet.ts.

Using the 2. element instead of the 6. it works and it works almost real time. The current result does not look as good as with resnet, but I will work on it.

**Comment by fangfufu** [comment]
btw, do you want to subscribe to https://github.com/webcamoid/webcamoid/issues/250? We are discussing similar issue over there.

**Comment by allo-** [comment]
I created a new repository with a clean rewrite of my script, as it does not need the simple_bodypix_python base.
The new version should work the same, but is in large parts rewritten to elimitate several redundant or ineffizient operations, so watch out for new bugs.

https://github.com/allo-/virtual_webcam_background
